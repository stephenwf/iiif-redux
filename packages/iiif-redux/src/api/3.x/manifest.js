import memoize from 'lodash.memoize';
import { createSelector } from 'reselect';
import mapById from '../../utility/mapById';
import * as technical from './iiif/technical';
import * as descriptive from './iiif/descriptive';
import * as linking from './iiif/linking';
import * as structural from './iiif/structural';
import {
  getAllAnnotationPages,
  getAllCanvases,
  getAllCanvasReferences,
  getAllContentResources,
  getAllRanges,
  getAllResources,
  getAllServices,
} from '../all';
import mapByIdOrId from '../../utility/mapByIdOrId';
import mapAllResources from '../../utility/mapAllResources';
import mapAllById from '../../utility/mapAllById';
import { standardFieldMappingFactory } from '../../../es/utility/new/standardFieldMappingFactory';

const manifest = memoize(selector => {
  /**
   * Technical properties
   *
   * - getId
   * - getType
   * - getViewingDirection
   * - getBehavior
   */
  const getId = createSelector(selector, technical.getId);

  const getType = createSelector(selector, technical.getType);

  const getViewingDirection = createSelector(
    selector,
    technical.getViewingDirection
  );

  const getBehavior = createSelector(selector, technical.getBehavior);

  /**
   * Descriptive Properties
   *
   * - getLabel
   * - getMetadata
   * - getSummary
   * - getThumbnail
   * - getPosterCanvas
   * - getRequiredStatement
   * - getRights
   * - getNavDate
   */
  const getLabel = createSelector(selector, descriptive.getLabel);

  const getMetadata = createSelector(selector, descriptive.getMetadata);

  const getSummary = createSelector(selector, descriptive.getSummary);

  const getThumbnailId = createSelector(selector, descriptive.getThumbnailId);

  const getThumbnail = createSelector(
    getThumbnailId,
    getAllContentResources,
    mapByIdOrId
  );

  const getPosterCanvasId = createSelector(
    selector,
    descriptive.getPosterCanvas
  );

  const getPosterCanvas = createSelector(
    getPosterCanvasId,
    getAllCanvases,
    mapById
  );

  const getRequiredStatement = createSelector(
    selector,
    descriptive.getRequiredStatement
  );

  const getRights = createSelector(selector, descriptive.getRights);

  const getNavDate = createSelector(selector, descriptive.getNavDate);

  /**
   * Linking properties
   *
   * - getSeeAlso
   * - getService
   * - getLogo
   * - getHomepage
   * - getRendering
   * - getPartOf
   * - getStart
   */
  const getSeeAlsoIds = createSelector(selector, linking.getSeeAlso);
  const getSeeAlso = createSelector(
    getSeeAlsoIds,
    getAllContentResources,
    mapAllById
  );

  const getServiceIds = createSelector(selector, linking.getService);
  const getService = createSelector(getServiceIds, getAllServices, mapAllById);

  const getLogoIds = createSelector(selector, linking.getLogo);
  const getLogo = createSelector(
    getLogoIds,
    getAllContentResources,
    mapAllById
  );

  const getHomepageId = createSelector(selector, linking.getHomepage);
  const getHomepage = createSelector(
    getHomepageId,
    getAllContentResources,
    mapById
  );

  const getRenderingIds = createSelector(selector, linking.getRendering);
  const getRendering = createSelector(
    getRenderingIds,
    getAllContentResources,
    mapAllById
  );

  const getPartOfId = createSelector(selector, linking.getPartOf);
  const getPartOf = createSelector(
    getPartOfId,
    getAllResources,
    mapAllResources
  );

  const getStartId = createSelector(selector, linking.getStart);
  const getStart = createSelector(
    getStartId,
    getAllCanvases,
    getAllCanvasReferences,
    (start, allCanvases, allCanvasReferences) => {
      if (!start) {
        return null;
      }
      if (start.schema === 'canvas') {
        return allCanvases[start.id];
      }
      return allCanvasReferences[start.id];
    }
  );

  /**
   * Structural Properties
   *
   * - getItems
   * - getStructures
   * - getAnnotations
   */
  const getItemIds = createSelector(selector, structural.getItems);
  const getItems = createSelector(getItemIds, getAllCanvases, mapAllById);

  const getCanvasIds = getItemIds;
  const getCanvases = getItems;

  const getStructureIds = createSelector(selector, structural.getStructures);
  const getStructures = createSelector(
    getStructureIds,
    getAllRanges,
    mapAllById
  );

  const getAnnotationIds = createSelector(selector, structural.getAnnotations);
  const getAnnotations = createSelector(
    getAnnotationIds,
    getAllAnnotationPages,
    mapAllById
  );

  return {
    getId,
    getType,
    getViewingDirection,
    getBehavior,
    getLabel,
    getMetadata,
    getSummary,
    getThumbnailId,
    getThumbnail,
    getPosterCanvasId,
    getPosterCanvas,
    getRequiredStatement,
    getRights,
    getNavDate,
    getSeeAlsoIds,
    getSeeAlso,
    getServiceIds,
    getService,
    getLogoIds,
    getLogo,
    getHomepageId,
    getHomepage,
    getRenderingIds,
    getRendering,
    getPartOfId,
    getPartOf,
    getStartId,
    getStart,
    getItemIds,
    getItems,
    getCanvasIds,
    getCanvases,
    getStructureIds,
    getStructures,
    getAnnotationIds,
    getAnnotations,
  };
});

export const mappings = standardFieldMappingFactory(manifest);

export default manifest;
