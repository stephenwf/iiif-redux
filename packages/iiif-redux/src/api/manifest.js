import presentation2Manifest from './2.x/manifest';
import { getAllAnnotationLists } from './all';
import resourceListSelectorFactory from '../utility/resourceListSelectorFactory';
import byIdSelectorFactory from '../utility/byIdSelectorFactory';

const manifest = presentation2Manifest;

export default manifest;

export const manifests = resourceListSelectorFactory(
  getAllAnnotationLists,
  manifest
);

export const manifestByIdSelector = byIdSelectorFactory(manifest, 'manifests');
