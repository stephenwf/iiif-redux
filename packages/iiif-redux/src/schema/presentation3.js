import { schema, normalize } from 'normalizr';
import { compose } from 'redux';
import addMissingIds from '../compat/addMissingIds';

function createEntity(name) {
  const options = {
    idAttribute: 'id',
  };

  return new schema.Entity(name, {}, options);
}

const collection = createEntity('collections'); // 1
const manifest = createEntity('manifests'); // 2
const canvas = createEntity('canvases'); // 3
const annotation = createEntity('annotations'); // 4
const annotationPage = createEntity('annotationPages'); // 5
const annotationCollection = createEntity('annotationCollection'); // 6
const range = createEntity('ranges'); // 7
const contentResource = createEntity('contentResources'); // 8
const choice = createEntity('choices'); // 9
const canvasReference = createEntity('canvasReference'); // 10

// Unofficial types.
const externalResource = createEntity('externalResources');
const service = createEntity('services');

// Union types
const partOf = new schema.Array(
  {
    collection,
    manifest,
    externalResource,
    annotationCollection,
  },
  entity => {
    switch (entity.type) {
      case 'Collection':
        return 'collection';
      case 'Manifest':
        return 'manifest';
      case 'AnnotationCollection':
        return 'annotationCollection';
      default:
        return 'externalResource';
    }
  }
);

const canvasOrReference = new schema.Union(
  {
    canvasReference,
    canvas,
  },
  entity => {
    if (entity.type === 'SpecificResource') {
      return 'canvasReference';
    }
    return 'canvas';
  }
);

const annotationBody = new schema.Union(
  {
    contentResource,
    externalResource,
    choice,
  },
  input => {
    switch (input.type) {
      case 'Application':
      case 'Dataset':
      case 'Image':
      case 'Sound':
      case 'Text':
      case 'Video':
      case 'TextualBody':
        return 'contentResource';
      case 'Choice':
        return 'choice';

      // @todo more content to add.
      default:
        return 'externalResource';
    }
  }
);

const rangeItem = new schema.Union(
  {
    canvas,
    range,
    canvasReference,
  },
  input => {
    switch (input.type) {
      case 'Canvas':
        return 'canvas';
      case 'Range':
        return 'range';
      case 'SpecificResource':
      default:
        // @todo default?
        return 'canvasReference';
    }
  }
);

// Root resource
/**
 * Dereferencable resources:
 *
 * - Collection (Required)
 * - Manifest (Required)
 * - Canvas (Recommended)
 * - Annotation (Recommended)
 * - AnnotationPage (Required)
 * - AnnotationCollection (Optional)
 * - Content Resource (Required)
 * - Range (Optional)
 */
const RESOURCE_TYPE_MAP = {
  Collection: 'collection',
  Sequence: 'sequence',
  Manifest: 'manifest',
  Canvas: 'canvas',
  AnnotationCollection: 'annotationCollection',
  AnnotationPage: 'annotationPage',
  Annotation: 'annotation',
  Range: 'range',
  // Content resources.
  Application: 'contentResource',
  Dataset: 'contentResource',
  Image: 'contentResource',
  Sound: 'contentResource',
  Text: 'contentResource',
  Video: 'contentResource',
  TextualBody: 'contentResource',
};
const resource = new schema.Union(
  {
    collection,
    manifest,
    canvas,
    annotationCollection,
    annotationPage,
    annotation,
    range,
    contentResource,
    service,
  },
  entity => {
    if (RESOURCE_TYPE_MAP[entity.type]) {
      return RESOURCE_TYPE_MAP[entity.type];
    }
    if (entity.profile) {
      return 'service';
    }
    throw Error('Unknown entity type');
  }
);

// ===========================================================================
// 1) Collections
// ===========================================================================
collection.define({
  // Structural
  items: new schema.Array(
    {
      collection,
      manifest,
    },
    entity => (entity.type === 'Manifest' ? 'manifest' : 'collection')
  ),
  annotations: [annotationPage],

  // Linking.
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  homepage: externalResource,
  rendering: [externalResource],
  partOf: partOf,

  // Extra.
  thumbnail: [contentResource],
  posterCanvas: canvas,
});

// ===========================================================================
// 2) Manifests
// ===========================================================================
manifest.define({
  items: [canvas],
  structures: [range],
  annotations: [annotationPage],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  homepage: externalResource,
  rendering: [externalResource],
  partOf: partOf,
  start: canvasOrReference,

  // Extra
  thumbnail: [contentResource],
  posterCanvas: canvas,
});

// ===========================================================================
// 3) Canvases
// ===========================================================================
canvas.define({
  // Structural.
  items: [annotationPage],
  annotations: [annotationPage],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  homepage: externalResource,
  rendering: [externalResource],
  partOf: partOf,

  // Extra
  thumbnail: [contentResource],
  posterCanvas: canvas,
});

// ===========================================================================
// 4) Annotation
// ===========================================================================
annotation.define({
  // Structural.
  body: annotationBody,
  target: canvas,

  // Linking
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  homepage: externalResource,
  rendering: [externalResource],
  partOf: partOf,

  // Extra
  thumbnail: [contentResource],
});

// ===========================================================================
// 5) Annotation Collection
// ===========================================================================
annotationCollection.define({
  // Structural.
  items: [annotationPage],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  homepage: externalResource,
  rendering: [externalResource],
  partOf: partOf,

  // extra
  thumbnail: [contentResource],

  // Paging.
  first: annotationPage,
  last: annotationPage,
});

// ===========================================================================
// 6) Annotation Page
// ===========================================================================
annotationPage.define({
  // Structural.
  items: [annotation],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  homepage: externalResource,
  rendering: [externalResource],
  partOf: partOf,

  // Extra
  thumbnail: [contentResource],

  // Paging.
  next: annotationPage,
  previous: annotationPage,
});

// ===========================================================================
// 7) Range
// ===========================================================================
range.define({
  // Structural.
  items: [rangeItem],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  homepage: externalResource,
  rendering: [externalResource],
  partOf: partOf,
  start: canvasOrReference,
  supplementary: annotationCollection,

  // Extra
  thumbnail: [contentResource],
});

// ===========================================================================
// 8) Content Resource
// ===========================================================================
contentResource.define({
  // Linking.
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  rendering: [externalResource],
  partOf: partOf,
});

// ===========================================================================
// 9) Choice
// ===========================================================================
choice.define({
  default: annotation,
  item: [annotation],
});

// ===========================================================================
// 10) Canvas Reference
// ===========================================================================
canvasReference.define({
  source: canvas,
});

const preprocess = compose(addMissingIds(3.0));

const normalizeResource = (rawResource, customSchema = resource) =>
  normalize(preprocess(rawResource), customSchema);

export {
  collection,
  manifest,
  canvas,
  annotation,
  annotationPage,
  annotationCollection,
  range,
  contentResource,
  choice,
  canvasReference,
  externalResource,
  service,
  partOf,
  annotationBody,
  rangeItem,
  RESOURCE_TYPE_MAP,
  resource,
  preprocess,
  normalizeResource as normalize,
};
