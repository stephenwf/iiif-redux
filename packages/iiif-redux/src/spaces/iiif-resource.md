# IIIF Resources
This state is where all of our entities get imported into. The actions here are dynamic
and will call off to other spaces with follow up actions.

## State structure
```js
const state = {
  dereferenced: {
    'http://.../collection.json': { 
      loading: true,
      requested: '00-00-0000:00:00:00', 
      ttl: 600,
      error: null, // or { message: 'Failed to fetch', ... }
    },
  },
  collections: {
    'http://.../collection.json': {/* ... */},
  },
  manifests: {},
  sequences: {},
  canvass: {},
  annotations: {},
  annotationLists: {},
  ranges: {},
  layers: {},
  imageResources: {},
  externalResources: {},
  services: {},
}
```

## Reusable actions
Requesting an item that can definitely be referenced (i.e. manifest):
```js
dispatch({
  type: 'REQUEST_IIIF_RESOURCE',
  payload: {
    resource: 'http://.../manifest.json',
    types: ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
    schema: schemas.manifest,
  }
});
```

Requesting an items fields, dereferencing if they don't exist.
```js
dispatch({
  type: 'REQUEST_IIIF_RESOURCE',
  payload: {
    resource: 'http://.../manifest.json',
    types: ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
    dereferenceIfMissing: ['sequences'],
    schema: schemas.imageService,
  }
});
```

Requesting a resource, with a TTL so if its requested again a fresh copy may be requested.
```js
dispatch({
  type: 'REQUEST_IIIF_RESOURCE',
  payload: {
    resource: 'http://.../manifest.json',
    types: ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
    ttl: 600,
    schema: schemas.manifest,
  }
});
```

Force-requesting an item, includes cache busting
```js
dispatch({
  type: 'REQUEST_IIIF_RESOURCE',
  payload: {
    resource: 'http://.../manifest.json',
    types: ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
    ttl: 600,
    forceFresh: true,
    schema: schemas.manifest,
  }
});
```

## Helper selectors
For use in containers for a more verbose way of showing resources current state while requesting.
```js
isResourceLoaded(id); // returns true/false, for loading screens.
didResourceError(id); // returns error object or null
```

## Potential sagas
In each of the spaces for the different resource types, we can have different reactions and
store different state. 

```js
function* manifestSaga() {
  yield takeEvery('MANIFEST_SUCCESS', handleSuccess);
} 

function* handleSuccess({ payload: { id } }) {
  // Set the current manifest to the one we just requested.
  yield put(setCurrentManifest(id));
  
  // Grab the state + start canvas
  const state = yield select();
  const startCanvas = getStartCanvas(state);

  // Set the canvas if its set.
  if (startCanvas) {
    yield put(setCurrentCanvasIndex(startCanvas)); // Another action creator.
  }
}
```
