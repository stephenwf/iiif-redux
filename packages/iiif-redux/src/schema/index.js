import * as presentation2 from './presentation2';
import * as presentation3 from './presentation3';

export { presentation2, presentation3 };

// Default is 2 until 3 is ratified?
export default presentation2;
