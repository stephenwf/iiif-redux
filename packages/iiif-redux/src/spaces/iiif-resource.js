import cachedFetch from 'fetch-unless-cached';
import { createActions, handleActions } from 'redux-actions';
import validUrl from 'valid-url';
import { call, put, select, takeEvery, all } from 'redux-saga/effects';
import deepmerge from 'deepmerge';
import update from 'immutability-helper';
import * as presentation2 from '../schema/presentation2';
import * as presentation3 from '../schema/presentation3';
import trackPresentationVersion from '../utility/trackPresentationVersion';
import {
  COLLECTION_ERROR,
  COLLECTION_REQUEST,
  COLLECTION_SUCCESS,
} from './collections';
import {
  MANIFEST_ERROR,
  MANIFEST_REQUEST,
  MANIFEST_SUCCESS,
} from './manifests';
import getPresentationVersionFromResource from '../utility/getPresentationVersionFromResource';
const debug = require('debug')('iiif-redux');

const IIIF_RESOURCE_REQUEST = 'IIIF_RESOURCE_REQUEST';
const IIIF_RESOURCE_SUCCESS = 'IIIF_RESOURCE_SUCCESS';
const IIIF_RESOURCE_ERROR = 'IIIF_RESOURCE_ERROR';
const IIIF_RESOURCE_REQUEST_UNKNOWN = 'IIIF_RESOURCE_REQUEST_UNKNOWN';
const IIIF_RESOURCE_IDENTIFY = 'IIIF_RESOURCE_IDENTIFY';

const DEFAULT_MAPPINGS = {
  collection: [COLLECTION_REQUEST, COLLECTION_SUCCESS, COLLECTION_ERROR],
  manifest: [MANIFEST_REQUEST, MANIFEST_SUCCESS, MANIFEST_ERROR],
};

const {
  iiifResourceRequestUnknown,
  iiifResourceRequest,
  iiifResourceSuccess,
  iiifResourceError,
} = createActions({
  [IIIF_RESOURCE_REQUEST_UNKNOWN]: (
    resourceId,
    options = {},
    mappings = DEFAULT_MAPPINGS
  ) => ({
    resourceId,
    dereferenceIfMissing: options.dereferenceIfMissing || false,
    ttl: options.ttl || 600,
    forceFresh: options.forceFresh,
    mappings,
  }),
  [IIIF_RESOURCE_REQUEST]: (resourceId, types, schema, options = {}) => ({
    resourceId,
    types,
    schema,
    dereferenceIfMissing: options.dereferenceIfMissing || false,
    ttl: options.ttl || 600,
    forceFresh: options.forceFresh,
  }),
  [IIIF_RESOURCE_SUCCESS]: (resourceId, normalizedResponse) => ({
    resourceId,
    normalizedResponse,
  }),
  [IIIF_RESOURCE_ERROR]: (resourceId, error) => ({
    resourceId,
    error,
  }),
});

const DEFAULT_STATE = {
  dereferenced: {},
  resources: {
    collections: {},
    sequences: {},
    manifests: {},
    canvases: {},
    canvasReferences: {},
    annotations: {},
    annotationLists: {},
    annotationPages: {},
    annotationCollections: {},
    ranges: {},
    layers: {},
    imageResources: {},
    externalResources: {},
    contentResources: {},
    choices: {},
  },
};

const resourceReducer = handleActions(
  {
    [iiifResourceSuccess]: (state, { payload: { normalizedResponse } }) =>
      deepmerge(state, normalizedResponse),
  },
  DEFAULT_STATE.resources
);

const dereferencedReducer = handleActions(
  {
    [iiifResourceRequest]: (state, { payload: { resourceId, ttl } }) =>
      update(state, {
        [resourceId]: {
          $set: {
            resourceId,
            ttl,
            requested: state[resourceId]
              ? state[resourceId].requested
              : new Date(),
            loading: state[resourceId] ? state[resourceId].loading : true,
          },
        },
      }),
    [iiifResourceRequestUnknown]: (state, { payload: { resourceId, ttl } }) =>
      update(state, {
        [resourceId]: {
          $set: {
            resourceId,
            ttl,
            loading: state[resourceId] ? state[resourceId].loading : true,
          },
        },
      }),

    [iiifResourceError]: (state, { payload: { resourceId, error } }) =>
      update(state, {
        [resourceId]: {
          loading: { $set: false },
          error: { $set: error },
        },
      }),

    [iiifResourceSuccess]: (state, { payload: { resourceId } }) =>
      update(state, {
        [resourceId]: {
          loading: { $set: false },
        },
      }),
  },
  DEFAULT_STATE.dereferenced
);

async function requestResource(resourceId, options) {
  /* istanbul ignore if */
  if (process.env.NODE_ENV === 'production') {
    return cachedFetch(resourceId);
  }

  return fetch(resourceId).then(resp => resp.json());
}

function* errorAction(type, resourceId, error) {
  yield put(iiifResourceError(resourceId, error));
  if (type) {
    yield put({ type, payload: { resourceId, error } });
  }
}

function* successAction(type, resourceId, normalizedResponse) {
  yield put(iiifResourceSuccess(resourceId, normalizedResponse));
  yield put({ type, payload: { resourceId, normalizedResponse } });
}

function* requestIiifResource({ payload }) {
  const { resourceId, types, ...options } = payload;
  if (types.length !== 3 && process.env.NODE_ENV !== 'production') {
    yield call(
      errorAction,
      null,
      resourceId,
      'DEV ERROR: You must pass in exactly 3 action types'
    );
  }

  const [REQUEST, SUCCESS, ERROR] = types;

  if (!validUrl.isWebUri(resourceId)) {
    yield call(
      errorAction,
      ERROR,
      resourceId,
      'Resource is not a valid Web URI.'
    );
    return;
  }

  const state = yield select();

  // @todo logic for cache.
  if (
    state.dereferenced[resourceId] &&
    state.dereferenced[resourceId].loading ===
      false /*&& options.forceFresh === false*/
  ) {
    debug('Skipping fetch for resource %s using cache.', resourceId);

    return;
  }

  yield put({ type: REQUEST, payload });

  try {
    debug('Fetching resource %s', resourceId);
    const response = yield call(requestResource, resourceId, options);

    const { result, entities } = yield call(
      importResource,
      response,
      resourceId
    );
    debug('Finished normalize resource %s', result.id);

    yield call(successAction, SUCCESS, result.id, entities);
  } catch (err) {
    debug('Error: %O', err);
    yield call(errorAction, ERROR, resourceId, err);
  }
}

function* importResource(response, resourceId) {
  const version = getPresentationVersionFromResource(response);
  if (version === 2) {
    return yield call(importPresentation2Resource, response, resourceId);
  }
  if (version === 3) {
    return yield call(importPresentation3Resource, response, resourceId);
  }

  throw new Error('Unknown IIIF Presentation version');
}

function* importPresentation2Resource(response, resourceId) {
  if (response['@id'] !== resourceId) {
    // @todo this will not catch "partOf" fields.
    debug(
      'Resource ID does not match requested presentation 2 resource, patching... Found: %s Expected: %s',
      response['@id'],
      resourceId
    );
    response['@id'] = resourceId;
  }

  debug('Starting normalize presentation 2 resource %s', resourceId);

  return trackPresentationVersion(
    2,
    presentation2.normalize(response, presentation2.resource)
  );
}

function* importPresentation3Resource(response, resourceId) {
  if (response.id !== resourceId) {
    // @todo this will not catch "partOf" fields.
    debug(
      'Resource ID does not match requested presentation 3 resource, patching... Found: %s Expected: %s',
      response.id,
      resourceId
    );
    response.id = resourceId;
  }

  debug('Starting normalize presentation 3 resource %s', resourceId);

  return trackPresentationVersion(
    3,
    presentation3.normalize(response, presentation3.resource)
  );
}

function* getResourceType(resourceId, response) {
  const presentationVersion = getPresentationVersionFromResource(response);
  if (presentationVersion === 2) {
    return yield call(getPresentation2ResourceType, resourceId, response);
  }
  if (presentationVersion === 3) {
    return yield call(getPresentation3ResourceType, resourceId, response);
  }

  throw new Error('Unknown IIIF Presentation version');
}

function* getPresentation2ResourceType(resourceId, response) {
  if (!presentation2.RESOURCE_TYPE_MAP[response['@type']]) {
    throw new Error('Unknown Presentation 2 resource type');
  }

  return presentation2.RESOURCE_TYPE_MAP[response['@type']];
}

function* getPresentation3ResourceType(resourceId, response) {
  if (!presentation3.RESOURCE_TYPE_MAP[response.type]) {
    throw new Error('Unknown Presentation 3 resource type');
  }

  return presentation3.RESOURCE_TYPE_MAP[response.type];
}

function* requestUnknownResource({
  payload: { resourceId, mappings, ...options },
}) {
  if (!validUrl.isWebUri(resourceId)) {
    yield put(iiifResourceError(resourceId, 'Resource is not a valid URL.'));
    return;
  }

  try {
    // 1) Request the resource.
    debug('Fetching unknown resource %s', resourceId);
    const response = yield call(requestResource, resourceId, options);
    if (!response) {
      yield put(iiifResourceError(resourceId, 'Unknown resource'));
      return;
    }

    const type = yield call(getResourceType, resourceId, response);

    if (!mappings[type]) {
      yield put(
        iiifResourceError(
          resourceId,
          `Resource type is not in configured mappings (${Object.keys(
            mappings
          ).join(', ')})`
        )
      );
      return;
    }

    // 3) Grab constants from a map (to be created).
    const [REQUEST, SUCCESS, ERROR] = mappings[type];

    // 4) Send to regular import (will be cached).
    yield put(iiifResourceRequest(resourceId, [REQUEST, SUCCESS, ERROR]));
  } catch (err) {
    yield put(iiifResourceError(resourceId, err.message));
  }
}

function* saga() {
  yield all([
    takeEvery(IIIF_RESOURCE_REQUEST, requestIiifResource),
    takeEvery(IIIF_RESOURCE_REQUEST_UNKNOWN, requestUnknownResource),
  ]);
}

export {
  saga,
  resourceReducer,
  dereferencedReducer,
  iiifResourceRequest,
  iiifResourceRequestUnknown,
  DEFAULT_STATE,
  IIIF_RESOURCE_REQUEST,
  IIIF_RESOURCE_SUCCESS,
  IIIF_RESOURCE_ERROR,
  IIIF_RESOURCE_REQUEST_UNKNOWN,
  IIIF_RESOURCE_IDENTIFY,
};
