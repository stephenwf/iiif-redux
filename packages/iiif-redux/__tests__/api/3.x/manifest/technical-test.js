import manifest from '../../../../src/api/3.x/manifest';
import { normalize } from '../../../../src/schema/presentation3';
import fixtures from './fixtures';
import { DEFAULT_STATE } from '../../../../src/spaces/iiif-resource';

describe('api/3.x/manifest/technical', () => {
  fixtures.forEach((fixture, num) => {
    describe(`Manifest number ${num}`, () => {
      const {
        result: { id },
        entities,
      } = normalize(fixture);
      const state = {
        resources: { ...DEFAULT_STATE.resources, ...entities },
      };
      const selectManifest = s => s.resources.manifests[id];

      const { getId, getType, getViewingDirection, getBehavior } = manifest(
        selectManifest
      );

      test('getId', () => {
        expect(getId(state)).toMatchSnapshot(`getId ${num}`);
      });

      test('getType', () => {
        expect(getType(state)).toMatchSnapshot(`getType ${num}`);
      });

      test('getViewingDirection', () => {
        expect(getViewingDirection(state)).toMatchSnapshot(
          `getViewingDirection ${num}`
        );
      });

      test('getBehavior', () => {
        expect(getBehavior(state)).toMatchSnapshot(`getBehavior ${num}`);
      });
    });
  });
});
