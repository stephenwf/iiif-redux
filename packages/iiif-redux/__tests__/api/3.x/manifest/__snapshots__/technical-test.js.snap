// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api/3.x/manifest/technical Manifest number 0 getBehavior: getBehavior 0 1`] = `
Array [
  "paged",
]
`;

exports[`api/3.x/manifest/technical Manifest number 0 getId: getId 0 1`] = `"https://example.org/iiif/book1/manifest"`;

exports[`api/3.x/manifest/technical Manifest number 0 getType: getType 0 1`] = `"Manifest"`;

exports[`api/3.x/manifest/technical Manifest number 0 getViewingDirection: getViewingDirection 0 1`] = `"right-to-left"`;

exports[`api/3.x/manifest/technical Manifest number 1 getBehavior: getBehavior 1 1`] = `undefined`;

exports[`api/3.x/manifest/technical Manifest number 1 getId: getId 1 1`] = `"https://example.org/iiif/collection/2"`;

exports[`api/3.x/manifest/technical Manifest number 1 getType: getType 1 1`] = `"Manifest"`;

exports[`api/3.x/manifest/technical Manifest number 1 getViewingDirection: getViewingDirection 1 1`] = `"left-to-right"`;

exports[`api/3.x/manifest/technical Manifest number 2 getBehavior: getBehavior 2 1`] = `undefined`;

exports[`api/3.x/manifest/technical Manifest number 2 getId: getId 2 1`] = `"https://example.org/iiif/collection/2"`;

exports[`api/3.x/manifest/technical Manifest number 2 getType: getType 2 1`] = `"Manifest"`;

exports[`api/3.x/manifest/technical Manifest number 2 getViewingDirection: getViewingDirection 2 1`] = `"left-to-right"`;
